dot/telemetry/mailer_test.go

```diff
@@ -22,32 +23,74 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-var upgrader = websocket.Upgrader{}
-var resultCh chan []byte
+func availablePort(t *testing.T) int {
+	t.Helper()
 
-func TestMain(m *testing.M) {
-	// start server to listen for websocket connections
-	upgrader.CheckOrigin = func(r *http.Request) bool { return true }
-	http.HandleFunc("/", listen)
-	go http.ListenAndServe("127.0.0.1:8001", nil)
+	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
+	require.NoError(t, err)
+
+	l, err := net.ListenTCP("tcp", addr)
+	require.NoError(t, err)
+	defer l.Close()
+
+	return l.Addr().(*net.TCPAddr).Port
+}
+
+func setupMockedListner(t *testing.T, upgrader *websocket.Upgrader, resCh chan<- []byte, logger log.LeveledLogger) func(w http.ResponseWriter, r *http.Request) {
+	t.Helper()
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		c, err := upgrader.Upgrade(w, r, nil)
+		require.NoError(t, err)
+
+		defer c.Close()
+
+		for {
+			_, msg, err := c.ReadMessage()
+			require.NoError(t, err)
+
+			resCh <- msg
+		}
```

Comment:
How does this for loop exit? Should we pass in a context to `setupMockListener` and exit on context cancellation?